prevent post-removal activity on range

tbg opened the pull request:
Scenario:
- client proposes command to raft, waits
- command commits, but hasn't gotten applied yet
- group gets removed, triggering the waiting client
- client decides the request has failed, leaves
- committed command executes
- r.pendingCmds still holds the command, so the contained trace
  panics (use-after-finalize).

The issue here is that `Replica`, `Store` and `Raft` are
moving parts, that can't fully move in lockstep with the removal.
(the store provides the Raft storage, so you can't remove the
group under the lock).

Instead, this change first "quiesces" the Replica, after which
it can be dropped from Raft, and, finally, from `Store`.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/2977)

<!-- Reviewable:end -->

tbg committed:
fix multiraft typo

tbg committed:
expose and use (net/trace).SetError()

tbg committed:
clarify a panic

bdarnell commented on storage/replica.go:
```diff
@@ -1558,3 +1557,23 @@ func (r *Replica) maybeAddToSplitQueue() {
+// In both cases, a RangeNotFoundError results.
+func (r *Replica) Quiesce() {
+	r.Lock()
+	r.proposeRaftCommandFn = func(_ cmdIDKey, _ roachpb.RaftCommand) <-chan error {
```
`proposeRaftCommandFn` was only meant to be used in testing; it's not properly synchronized for this usage (it's read without a lock)

tschottdorf replied:
changed using the bool field.

bdarnell commented on storage/replica.go:
```diff
@@ -1558,3 +1557,23 @@ func (r *Replica) maybeAddToSplitQueue() {
+	for id, cmd := range r.pendingCmds {
+		cmd.done <- roachpb.ResponseWithError{Err: multiraft.ErrGroupDeleted}
+		// Deleting while iterating is ok in Go. We don't just want to `nil`
+		// this since there could be commands being queued up for execution on
```
But don't you want all commands which might be added to `pendingCmds` to fail? It looks somewhat fragile that we fail all currently-pending commands while allowing others to be written (with the expectation that they will fail when they reach the patched `proposeRaftCmdFn`. I think it would be better to add an explicit `quiesced` bool field to the `Replica` and use it in `proposeRaftCommand` (and then we should be able to simply nil out the `pendingCmds` map)

tschottdorf replied:
I've added a `quiesced` field. Always a bit hesitant to introduce new fields but I suppose it's too sketchy to encode this information as `pendingCmds = nil` (the two are equal)?

bdarnell replied:
Yeah, I think it's better to add the explicit field even if it happens to be equivalent to `pendingCmds == nil` for now. It seems less likely to break with future changes and it's easier to grep for.

tschottdorf commented:
updated and added a test, PTAL.

bdarnell commented on storage/replica.go:
```diff
@@ -862,14 +865,22 @@ func (r *Replica) proposeRaftCommand(ctx context.Context, ba roachpb.BatchReques
 	var errChan <-chan error
-	if r.proposeRaftCommandFn != nil {
-		errChan = r.proposeRaftCommandFn(idKey, raftCmd)
+	if r.quiesced {
```
The lock needs to be held while accessing `r.quiesced` (I'm surprised and worried that the race detector didn't find this).

tschottdorf replied:
Clearly not awake yet. We don't have many tests that actually `Quiesce` (and the one I wrote is synchronized so that the race can't happen).

bdarnell replied:
We don't have many, but a lot of the ones in `client_raft_test.go` do. But I guess we don't have any tests that unreplicate while something else is going on.

bdarnell commented:
LGTM except for the locking.

tbg committed:
prevent post-removal activity on range

as observed in #2593, the following could happen:

* client proposes command to raft, waits
* command commits, but hasn't reached client yet
* group gets removed, triggering the waiting client
* client decides the request is done, leaves
* committed command executes
* r.pendingCmds still holds the command, so the contained trace
  panics (use-after-finalize).

The issue here is that `Replica`, `Store` and `Raft` are
moving parts, that can't fully move in lockstep with the removal.
(the store provides the Raft storage, so you can't remove the
group under the lock).

Instead, this change first "quiesces" the Replica, after which
it can be dropped from Raft, and, finally, from `Store`.

mrtracy commented:
LGTM
<!-- Reviewable comment -K28X5ynXS0-QAIVOb0P:676834233 -->

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2977)
<!-- Sent from Reviewable.io -->
